include/fiction/networks/virtual_pi_network.hpp

```diff
@@ -0,0 +1,693 @@
+    }
+
+    /**
+     * @brief Reimplementation of num_gates(). Retrieves the number of gates in the virtual_pi_network.
+     *
+     * @return The number of gates in the virtual_pi_network.
+     */
+    auto num_gates() const
+    {
+        return static_cast<uint32_t>(_storage->nodes.size() - virtual_inputs->size() - _storage->inputs.size() - 2);
+    }
+
+    /**
+     * @brief Reimplementation of size(). Get the size of the virtual_pi_network.
+     *
+     * @return The size of the virtual_pi_network as a uint32_t.
+     */
+    auto size() const
+    {
+        return static_cast<uint32_t>(_storage->nodes.size());
+    }
+
+    /**
+     * \brief Calculate the real size of the virtual_pi_network.
+     *
+     * This function calculates the real size of the virtual_pi_network by subtracting the size of the virtual_inputs
+     * vector from the size of the _storage->nodes vector.
+     *
+     * \return The real size of the virtual_pi_network as a uint32_t.
+     */
+    [[nodiscard]] auto size_real() const
+    {
+        return static_cast<uint32_t>(_storage->nodes.size() - virtual_inputs->size());
+    }
+
+    /**
+     * @brief Create a virtual PI mapping to a real PI.
+     *
+     * This function creates a virtual PI mapping to a real PI in the network. It adds a new node to the
+     * `_storage->nodes` vector and inserts the index of the new node into the `virtual _inputs` vector. It also updates
+     * the data of the new node.
+     *
+     * @param real_pi The node representing the real PI in the network.
+     * @return The index of the newly created virtual PI.
+     */
+    signal create_virtual_pi(const node& real_pi)
+    {
+        const auto index = _storage->nodes.size();
+        _storage->nodes.emplace_back();
+        virtual_inputs->emplace_back(index);
+        _storage->nodes[index].data[1].h1 = 2;
+        map.emplace_back(index, real_pi);
+        return index;
+    }
+
+    /**
+     * @brief Reimplementation of is_pi(). Check if a given node is a primary input (PI) in the virtual_pi_network.
+     *
+     * A node is considered a PI if it exists in the virtual_inputs vector or the inputs vector of the _storage object.
+     *
+     * @param n The node to check.
+     * @return True if the node is a PI, false otherwise.
+     */
+    [[nodiscard]] bool is_pi(node const& n) const
+    {
+        return std::find(virtual_inputs->begin(), virtual_inputs->end(), n) != virtual_inputs->end() ||
+               std::find(_storage->inputs.begin(), _storage->inputs.end(), n) != _storage->inputs.end();
+    }
+
+    /**
+     * @brief Check if a given node is a virtual primary input (PI) in the virtual_pi_network.
+     *
+     * A node is considered a virtual PI if it exists in the virtual_inputs vector of the virtual_pi_network.
+     *
+     * @param n The node to check.
+     * @return True if the node is a virtual PI, false otherwise.
+     */
+    [[nodiscard]] bool is_pi_virtual(node const& n) const
```

Comment:
```suggestion
    [[nodiscard]] bool is_virtual_pi(node const& n) const
```
Because we cannot limit the input to PIs, this function checks for both the node being a PI and it being real, thus the proposed name is more accurate.